run-regression-test: add notification support by E-Mail

kenhys opened the pull request:
There is a case that groonga-query-log is invoked from shell script and it needs to report test result
as a use-case.
If groonga-query-log supports such notification feature by default, it makes more usable. (it also works easily on windows)

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
         @read_timeout = Groonga::Client::Default::READ_TIMEOUT
+
+        @mail_notification = false
+        @mail_subject_on_succeeded = "Succeeded"
```
We can use `success`.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
+
+        @mail_notification = false
+        @mail_subject_on_succeeded = "Succeeded"
+        @mail_subject_on_failure = "Failed"
```
We can use `failure`.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
+        @mail_notification = false
+        @mail_subject_on_succeeded = "Succeeded"
+        @mail_subject_on_failure = "Failed"
+        @mail_from = "groonga-query-log@example.com"
```
`"groonga-query-log@#{Socket.hostname}` is better as the default from. Because `example.com` doesn't exist.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
+        @mail_subject_on_succeeded = "Succeeded"
+        @mail_subject_on_failure = "Failed"
+        @mail_from = "groonga-query-log@example.com"
+        @mail_to = "noreply@localhost"
```
It's not a good default value. We must require the value. `--mail-notification` may not needed.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
+        @mail_from = "groonga-query-log@example.com"
+        @mail_to = "noreply@localhost"
+        @smtp_server = "localhost"
+        @smtp_auth_user = "unknown"
```
`nil` (no SMTP auth) is the suitable default value.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
+        @mail_to = "noreply@localhost"
+        @smtp_server = "localhost"
+        @smtp_auth_user = "unknown"
+        @smtp_auth_password = "unknown"
```
`nil` (no SMTP auth) is the suitable default value.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
+        @smtp_auth_user = "unknown"
+        @smtp_auth_password = "unknown"
+
+        @start_time = Time.now
```
It should execute just before `tester.run`. If we execute this here, we count needless time.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -70,7 +85,14 @@ def run(command_line)
                             new_groonga_server,
                             tester_options)
-        tester.run
+        tester_result = tester.run
```
`success` is better name.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -70,7 +85,14 @@ def run(command_line)
+
+        if notifier_options[:mail_notification]
+          notifier = EMailNotifier.new(tester_result, @start_time, notifier_options)
+          notifier.run
```
`notify` is better name.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -258,6 +317,20 @@ def tester_options
+          :mail_from => @mail_from,
+          :mail_to => @mail_to
+        }
+        directory_options.merge(options)
```
Is this needed?

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -258,6 +317,20 @@ def tester_options
+          :smtp_server => @smtp_server,
+          :smtp_auth_user => @smtp_auth_user,
+          :smtp_auth_password => @smtp_auth_password,
+          :mail_subject_on_failure => @mail_subject_on_failure,
```
Is `mail_` prefix needed?

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +644,75 @@ def use_persistent_cache?
+          elapsed_minutes = elapsed / 60 % 60
+          elapsed_hours = elapsed / 60 / 60 % 24
+          elapsed_days = elapsed / 60 / 60 / 24
+          sprintf("Elapsed: %ddays %02d:%02d:%02d\n",
```
`"..." % [...]` is better than `sprintf` in Ruby.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -70,7 +85,14 @@ def run(command_line)
+        tester_result = tester.run
+
+        if notifier_options[:mail_notification]
+          notifier = EMailNotifier.new(tester_result, @start_time, notifier_options)
```
Passing elapsed time instead of start time is better. If we pass elapsed time measurement is a responsibility of this class not `EMailNotifier`.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +644,75 @@ def use_persistent_cache?
         end
       end
+
+      class EMailNotifier
```
We should use only `email` or `mail` for consistency.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +644,75 @@ def use_persistent_cache?
+
+        def run
+          $stdout = Tempfile.open("format_stdout")
+          command = GroongaQueryLog::Command::FormatRegressionTestLogs.new
```
We should not swap `$stdout`. We should extract a class from `FormatRegressionTestLogs` and the extracted class should accept the output.

kou replied:
It should be done in another pull request before this pull request.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +644,75 @@ def use_persistent_cache?
+          $stdout = STDOUT
+          subject = @options[:mail_subject_on_succeeded]
+          content = elapsed_time_format
+          unless @tester_result
```
We should not use `unless ... else ...`. It's confused.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +644,75 @@ def use_persistent_cache?
+
+          mail = <<-MAIL.gsub(/\r?\n/, "\r\n")
+#{header}
+#{body}
```
Need an empty line between header and body.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +644,75 @@ def use_persistent_cache?
+#{header}
+#{body}
+          MAIL
+          smtp = Net::SMTP.new(@options[:smtp_server], 587)
```
Port must be customizable.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +644,75 @@ def use_persistent_cache?
+#{body}
+          MAIL
+          smtp = Net::SMTP.new(@options[:smtp_server], 587)
+          smtp.enable_starttls
```
StartTLS must be customizable.

kou commented:
Ah, @komainu8 should have reviewed not me...

kenhys commented:
except for swapping  $stdout, fixes for review comment are committed.
#39 shoud be merged before because it solves without swapping $stdout.

komainu8 commented:
LGTM :+1:

kenhys commented:
As #39 is merged, I've fixed #39 related code and rebase & squashed commits.

kenhys commented:
I've fixed based on review feedbacks again except MailNotifier.notify should accept `StringIO`.
I can't guess the benefit of it yet.

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -56,6 +60,17 @@ def initialize
+        @smtp_auth_password = nil
+        @smtp_starttls = false
+        @smtp_port = 25
+
```
Could you remove this empty line?

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -571,6 +650,79 @@ def use_persistent_cache?
       end
+
+      class MailNotifier
+
```
Could you remove this empty line?

kou commented:
> I've fixed based on review feedbacks again except `MailNotifier.notify` should accept `StringIO`.
> I can't guess the benefit of it yet.

We can avoid IO related error such as permission denied and disk full.
`StringIO` is faster than `Tempfile`. It's helpful on test.

kenhys commented:
>>  I've fixed based on review feedbacks again except MailNotifier.notify should accept StringIO.
    I can't guess the benefit of it yet.

> We can avoid IO related error such as permission denied and disk full.
StringIO is faster than Tempfile. It's helpful on test.

@kou 
Thank you for explanation, I've got it and fixed to eliminate Tempfile and remove some other unused code. Could you review again, please?

kou commented on lib/groonga-query-log/command/run-regression-test.rb:
```diff
@@ -585,6 +663,74 @@ def use_persistent_cache?
+          @path = @options[:path] || "results"
+        end
+
+        def notify(output=nil)
```
The `output` option is awkward.
